From 54e6dea31166f1c5e2b81c8b629801a66824d19c Mon Sep 17 00:00:00 2001
From: jhonboy121 <alfredmathew05@gmail.com>
Date: Fri, 18 Mar 2022 16:09:21 +0530
Subject: [PATCH] BluetoothControllerImpl: Synchronize mConnectedDevices access

in getBatteryLevel()

03-17 21:45:13.643  2245  2245 E AndroidRuntime: FATAL EXCEPTION: main
03-17 21:45:13.643  2245  2245 E AndroidRuntime: Process: com.android.systemui, PID: 2245
03-17 21:45:13.643  2245  2245 E AndroidRuntime: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at java.util.ArrayList.get(ArrayList.java:437)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.systemui.statusbar.policy.BluetoothControllerImpl.getBatteryLevel(BluetoothControllerImpl.java:329)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.systemui.statusbar.policy.BluetoothControllerImpl.updateBattery(BluetoothControllerImpl.java:335)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.systemui.statusbar.policy.BluetoothControllerImpl.updateConnected(BluetoothControllerImpl.java:282)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.systemui.statusbar.policy.BluetoothControllerImpl.onDeviceAttributesChanged(BluetoothControllerImpl.java:379)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.settingslib.bluetooth.CachedBluetoothDevice.dispatchAttributesChanged(CachedBluetoothDevice.java:904)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.settingslib.bluetooth.CachedBluetoothDevice.lambda$refresh$0(CachedBluetoothDevice.java:612)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.settingslib.bluetooth.CachedBluetoothDevice.$r8$lambda$4IYXIWE9azg2QshMjGh5evIGUmg(Unknown Source:0)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.settingslib.bluetooth.CachedBluetoothDevice$$ExternalSyntheticLambda1.run(Unknown Source:2)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at android.os.Handler.handleCallback(Handler.java:938)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at android.os.Handler.dispatchMessage(Handler.java:99)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at android.os.Looper.loopOnce(Looper.java:201)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at android.os.Looper.loop(Looper.java:288)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at android.app.ActivityThread.main(ActivityThread.java:7871)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at java.lang.reflect.Method.invoke(Native Method)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550)
03-17 21:45:13.643  2245  2245 E AndroidRuntime:	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)

Change-Id: I036ffeb434ddbd7f2abfcfe950414319458f455b
Signed-off-by: Adithya R <gh0strider.2k18.reborn@gmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../systemui/statusbar/policy/BluetoothControllerImpl.java  | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BluetoothControllerImpl.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BluetoothControllerImpl.java
index 6775be951dab..b705887be01e 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BluetoothControllerImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BluetoothControllerImpl.java
@@ -334,8 +334,10 @@ public class BluetoothControllerImpl implements BluetoothController, BluetoothCa
 
     @Override
     public int getBatteryLevel() {
-        if (!mConnectedDevices.isEmpty()) {
-            return mConnectedDevices.get(0).getBatteryLevel();
+        synchronized (mConnectedDevices) {
+            if (!mConnectedDevices.isEmpty()) {
+                return mConnectedDevices.get(0).getBatteryLevel();
+            }
         }
         return -1;
     }
-- 
2.49.0

